### ISSUE TITLE
An Option to stop - 'rewrite THIS to UNDEFINED' 

### ISSUE THREAD

**Issue by jaguart** [rollup/rollup#3788]
### Feature Use Case

TypeScript, since v3.8 has provided a shim to support using ES2020 hard-privates in ES6/ES2015.
Hard-privates are a desirable feature for writing modern Javascript.

When targeting ES6/ES2015, TypeScript emits:
```javascript
var __classPrivateFieldGet=this&&this.__classPrivateFieldGet||function(e,i){...
var __classPrivateFieldSet=this&&this.__classPrivateFieldSet||function(e,i,t){
```
This is a cheap run-time efficiency, to create a global function rather than having this identical function in every class that has hard-privates, dependant on run-time handling of *this* in a module outside a class.

Rollup unconditionally rewrites this to:
```javascript
var __classPrivateFieldGet=undefined&&undefined.__classPrivateFieldGet||function(e,i){...
var __classPrivateFieldSet=undefined&&undefined.__classPrivateFieldSet||function(e,i,t){
```
I know that in the ES6/ES2015 module spec *this* does not have a value outside a class. In reality, Rollup cannot know the details of all potential run-time environments where the script might be executed.

### Feature Proposal

A new option that disables this specific rewrite.

Perhaps something like:
```
  output: {
    rewriteThis: false
  }
```
command-line: ``--no-rewrite-this``

or something that fits the Rollup style.

#### Alternate

The rewrite code could be tweaked to ignore the specific pattern `this && this` - as the emitter has clearly gone to the trouble of testing the existence of *this*.

**Comment by lukastaegert** [comment]
For starters, use this option: https://rollupjs.org/guide/en/#context

Assuming that `this` is the global object is unfortunately wrong for some environments as well, and Rollup specifically runs under the assumption we are bundling ES modules where `this` is `undefined`. For that reason, the commonjs plugin will actually take care of rewriting `this` to a reference to the global object inside CommonJS files to match Node's `this` value.

**Comment by lukastaegert** [comment]
So as a matter of fact, `context: "this"` is the `--no-rewrite-this` option you suggested.

**Comment by jaguart** [comment]
Doh - thank you 👍 

Wouldn't it be better, when format `es` just to rewrite it to `globalThis`  for both browser and node?

**Comment by lukastaegert** [comment]
As I said, the assumption is that anything that is processed by Rollup core, i.e. after plugins have transformed it, is a proper ES module. That implies "strict" mode which in turn implies that global `this` is `undefined`. This is about spec compliance and any diverging behaviour should be handled in plugins. If you have an ES module as input that assumes that global `this` is the global object, then this module is broken according to the ES module specs. It will not work if you have it in a `type: module` script tag in the browser or import it as an ES module in Node 14.

So to sum it up, if TypeScript uses this transform when targeting the ES module format and assumes that `this` is the global object, then the TypeScript transform is invalid and an issue should be opened there.

**Comment by lukastaegert** [comment]
Ah, I was wrong about strict mode, that one only changes the `this` value in functions. The global `this` value is a consequence of the ES module format. But yes, I can confirm that the TypeScript transform is broken, in this example you can clearly see that they are combining an `export`, that only works in an ES module, with a call to global `this`: https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwDYEMDOa4DEITgbwCg4S4BiAWwE8AFKASwDcUZgB+ALjgDsBXCgEbAoAbkLFSFYDAAWEACYAKAJQEJpUrPpoAdJVoNmrOAF44ARjEaNCCNzQQkwHUggBzRVt366TFsGUrUgBfQlCSIA

**Comment by lukastaegert** [comment]
Ok, broken is probably too harsh, it is just unnecessary to check `this` at all, so they could improve here to avoid the warning in Rollup. My recommendation for projects would be to use `importHelpers` anyway, that likely solves the issue: https://www.typescriptlang.org/play?importHelpers=true#code/KYDwDg9gTgLgBAYwDYEMDOa4DEITgbwCg4S4BiAWwE8AFKASwDcUZgB+ALjgDsBXCgEbAoAbkLFSFYDAAWEACYAKAJQEJpUrPpoAdJVoNmrOAF44ARjEaNCCNzQQkwHUggBzRVt366TFsGUrUgBfQlCSIA

**Comment by jaguart** [comment]
Fab - thanks for the detailed responses, and recommendations - I agree these resolve the issue.
